zetaclient/evm/base_transaction_data.go

```diff
@@ -0,0 +1,179 @@
+package evm
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"math/big"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/rs/zerolog"
+	"github.com/zeta-chain/zetacore/common"
+	"github.com/zeta-chain/zetacore/x/crosschain/types"
+	"github.com/zeta-chain/zetacore/zetaclient/interfaces"
+)
+
+const (
+	MinGasLimit = 100_000
+	MaxGasLimit = 1_000_000
+)
+
+// BaseTransactionData is a data structure containing input fields used to construct each type of transaction.
+// This is populated using cctx and other input parameters passed to TryProcessOutTx
+type BaseTransactionData struct {
+	srcChainID *big.Int
+	toChainID  *big.Int
+	sender     ethcommon.Address
+	to         ethcommon.Address
+	asset      ethcommon.Address
+	amount     *big.Int
+	gasPrice   *big.Int
+	gasLimit   uint64
+	message    []byte
+	nonce      uint64
+	height     uint64
+
+	// sendHash field is the inbound message digest that is sent to the destination contract
+	sendHash [32]byte
+
+	// outboundParams field contains data detailing the receiver chain and outbound transaction
+	outboundParams *types.OutboundTxParams
+}
+
+// SetChainAndSender populates the destination address and Chain ID based on the status of the cross chain tx
+// returns true if transaction should be skipped
+// returns false otherwise
+func (txData *BaseTransactionData) SetChainAndSender(cctx *types.CrossChainTx, logger zerolog.Logger) bool {
+	switch cctx.CctxStatus.Status {
+	case types.CctxStatus_PendingRevert:
+		txData.to = ethcommon.HexToAddress(cctx.InboundTxParams.Sender)
+		txData.toChainID = big.NewInt(cctx.InboundTxParams.SenderChainId)
+		logger.Info().Msgf("Abort: reverting inbound")
+	case types.CctxStatus_PendingOutbound:
+		txData.to = ethcommon.HexToAddress(cctx.GetCurrentOutTxParam().Receiver)
+		txData.toChainID = big.NewInt(cctx.GetCurrentOutTxParam().ReceiverChainId)
+	default:
+		logger.Info().Msgf("Transaction doesn't need to be processed status: %d", cctx.CctxStatus.Status)
+		return true
+	}
+	return false
+}
+
+// SetupGas sets the gas limit and price
+func (txData *BaseTransactionData) SetupGas(
+	cctx *types.CrossChainTx,
+	logger zerolog.Logger,
+	client interfaces.EVMRPCClient,
+	chain *common.Chain,
```

Comment:
I think we can just use a non-pointer parameter here?

Reply:
This function returns a pointer common.GetChainFromChainID, so I decided to maintain the pointer to avoid conversions or dereferencing.